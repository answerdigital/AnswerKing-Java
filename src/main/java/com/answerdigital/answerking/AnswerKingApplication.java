package com.answerdigital.answerking;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class AnswerKingApplication {
    public static void main(final String[] args) {
	SpringApplication.run(AnswerKingApplication.class, args);
    }
}
