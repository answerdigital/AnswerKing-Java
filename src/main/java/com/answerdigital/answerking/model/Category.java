package com.answerdigital.answerking.model;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * The class Category represents a group of Products with similar traits.
 */
@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED, force = true)
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Pattern(regexp = "^[a-zA-Z\s-]*", message = "Category name must only contain letters, spaces and dashes")
    private String name;

    @NotBlank
    @Pattern(regexp = "^[a-zA-Z\s.,!?0-9-']*",
            message = "Category description can only contain letters, numbers, spaces and !?-.,' punctuation")
    private String description;

    @CreationTimestamp
    @Column(name = "created_on")
    private LocalDateTime createdOn;

    @UpdateTimestamp
    @Column(name = "last_updated")
    private LocalDateTime lastUpdated;

    private boolean retired;

    @OneToMany(mappedBy="category", fetch = FetchType.EAGER)
    private List<Product> products = new ArrayList<>();

    /**
     * This method adds a Product {@link com.answerdigital.answerking.model.Product}
     * to the Category.
     */
    public void addProduct(final Product product) {
        products.add(product);
    }

    /**
     * This method removes a Product {@link com.answerdigital.answerking.model.Product}
     * from the Category.
     */
    public void removeProduct(final Product product) {
        products.remove(product);
    }
}
