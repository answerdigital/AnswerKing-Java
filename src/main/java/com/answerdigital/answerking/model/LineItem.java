package com.answerdigital.answerking.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.Objects;

@Entity
@Table(name = "order_product")
@Getter
@Setter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED, force = true)
@AllArgsConstructor
public class LineItem {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    @JoinColumn(name = "order_id")
    private Order order;

    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    @JoinColumn(name = "product_id")
    private Product product;

    @Column(name = "quantity")
    private Integer quantity;

    public LineItem(final Order order, final Product product, final int quantity) {
        this.order = order;
        this.product = product;
        this.quantity = quantity;
    }

    /**
     * This method checks if the LineItem is the exact same as the selected object
     * (normally Product {@link com.answerdigital.answerking.model.LineItem} object).
     * @return A boolean verifying whether the objects are the same or not.
     */
    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final LineItem lineItem = (LineItem) o;
        return (
            Objects.equals(order, lineItem.order)
            && Objects.equals(product, lineItem.product)
            && Objects.equals(quantity, lineItem.quantity));
    }

    /**
     * This method returns a hash of the LineItem.
     * @return A Hash value for an Order ID, Product ID and Product Quantity.
     */
    @Override
    public int hashCode() {
        return Objects.hash(order, product, quantity);
    }

    @Override
    public String toString() {
        return "LineItem{" +
                "id=" + id +
                ", orderId=" + order.getId() +
                ", productId=" + product.getId() +
                ", quantity=" + quantity +
                '}';
    }
}
