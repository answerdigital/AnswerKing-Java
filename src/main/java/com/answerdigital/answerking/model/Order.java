package com.answerdigital.answerking.model;

import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import lombok.AllArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @CreationTimestamp
    private LocalDateTime createdOn;

    @UpdateTimestamp
    private LocalDateTime lastUpdated;

    @Enumerated(EnumType.STRING)
    private OrderStatus orderStatus;

    @OneToMany(
            mappedBy = "order",
            cascade = CascadeType.ALL,
            orphanRemoval = true,
            fetch = FetchType.EAGER
    )
    private Set<LineItem> lineItems = new HashSet<>();

    public Order() {
        this.orderStatus = OrderStatus.CREATED;
    }

    public void addLineItem(final LineItem lineItem) {
        lineItems.add(lineItem);
    }

    public void clearLineItems() {
        lineItems.clear();
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

}
