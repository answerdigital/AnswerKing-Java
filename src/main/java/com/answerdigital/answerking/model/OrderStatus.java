package com.answerdigital.answerking.model;

/**
 * The class OrderStatus represents the current status of an
 * Order {@link com.answerdigital.answerking.model.Order}.
 */
public enum OrderStatus {
    CREATED,
    COMPLETE,
    CANCELLED
}
