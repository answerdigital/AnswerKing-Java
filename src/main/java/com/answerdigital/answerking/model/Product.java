package com.answerdigital.answerking.model;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PreRemove;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "product")
@Getter
@Setter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED, force = true)
@AllArgsConstructor
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private String description;

    @Column(precision = 12, scale = 2)
    private BigDecimal price;

    private boolean retired;

    @ManyToOne
    @JoinColumn(name="category_id", nullable=false)
    private Category category;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    private Set<LineItem> lineItems = new HashSet<>();

    @ManyToMany(mappedBy = "products", cascade = CascadeType.MERGE, fetch = FetchType.EAGER)
    private Set<Tag> tags = new HashSet<>();

    @PreRemove
    private void removeProductsFromCategories() {
        category.removeProduct(this);
    }

    public BigDecimal getPrice() {
        return price.setScale(2, RoundingMode.DOWN);
    }
}
