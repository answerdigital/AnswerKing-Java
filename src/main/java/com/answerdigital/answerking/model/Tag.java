package com.answerdigital.answerking.model;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.Builder;

import java.util.List;

@Entity
@Table(name = "tag")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Tag {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private String description;

    private boolean retired;

    @ManyToMany(cascade = CascadeType.MERGE, fetch = FetchType.EAGER)
    @JoinTable(
            name = "tag_product",
            joinColumns = { @JoinColumn(referencedColumnName = "id") },
            inverseJoinColumns = { @JoinColumn(referencedColumnName = "id") }
    )
    private List<Product> products;

}
