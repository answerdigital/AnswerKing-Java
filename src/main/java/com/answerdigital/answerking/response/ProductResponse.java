package com.answerdigital.answerking.response;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;

/**
 * The class ProductResponse represents the Product {@link com.answerdigital.answerking.model.Product}
 * to be returned, after the end-user has sent an API endpoint request.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ProductResponse {

    private Long id;

    private String name;

    private String description;

    private BigDecimal price;

    private Long category;

    private boolean retired;
}
