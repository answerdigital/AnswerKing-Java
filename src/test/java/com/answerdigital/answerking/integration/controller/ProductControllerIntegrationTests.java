package com.answerdigital.answerking.integration.controller;

import com.answerdigital.answerking.AnswerKingApplication;
import com.answerdigital.answerking.model.Category;
import com.answerdigital.answerking.model.Product;
import com.answerdigital.answerking.response.ProductResponse;
import com.answerdigital.answerking.utility.AbstractContainerBaseTest;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.web.servlet.MockMvc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest(classes = {AnswerKingApplication.class, Category.class, Product.class, ProductResponse.class},
                webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc(addFilters = false)
@ActiveProfiles("integration-test")
@ExtendWith({SnapshotExtension.class})
class ProductControllerIntegrationTests extends AbstractContainerBaseTest {

    private Expect expect;

    public String uuidRegex = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}";

    @Autowired
    private MockMvc mvc;

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getAllProductsTest() throws Exception {

        final var response = mvc.perform(get("/products"))
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response);
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql"})
    void getAllProductsEmptyTest() throws Exception {

        final var response = mvc.perform(get("/products"))
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        assertEquals("[]", response);
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getProductByIDTest() throws Exception {

        final var response = mvc.perform(get("/products/1")) // Need to get ID from seed really
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response);
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getProductByIDNotFoundTest() throws Exception {

        final var response = mvc.perform(get("/products/{id}", 10000L))
            .andExpect(status().isNotFound())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getProductByIDInvalidTest() throws Exception {

        final var response = mvc.perform(get("/products/invalid")) 
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }

    @Test
    @Tag("POST")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void postProductTest() throws Exception {

        final var response = mvc.perform(post("/products")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{\"name\": \"Kabab\", \"description\": \"Product 1 desc\", \"price\": 10.34, \"categoryId\": 1}")) 
            // Need to add static fixtures instead of writing this in tests
            .andExpect(status().isCreated())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response); 
    }

    @Test
    @Tag("POST")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void postProductTestInvalidBody() throws Exception {

        final var response = mvc.perform(post("/products")
            .contentType(MediaType.APPLICATION_JSON)
            .content("This is not valid JSON")) 
            // Need to add static fixtures and be data driven to check all different ways body can be incorrect
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }

    @Test
    @Tag("PUT")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void putProductTest() throws Exception {

        final var response = mvc.perform(put("/products/2")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{\"name\": \"Kabab\", \"description\": \"Product 1 desc\", \"price\": 10.34, \"categoryId\": 1}")) 
            // Need to add static fixtures instead of writing this in tests
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response); 
    }

    @Test
    @Tag("PUT")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void putProductTestInvalidBody() throws Exception {

        final var response = mvc.perform(put("/products/{id}", 2L)
            .contentType(MediaType.APPLICATION_JSON)
            .content("This is not valid JSON")) 
            // Need to add static fixtures and be data driven to check all different ways body can be incorrect
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }

    @Test
    @Tag("PUT")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void putProductByIDInvalidTest() throws Exception {

        final var response = mvc.perform(put("/products/{id}", "invalid")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{\"name\": \"Kabab\", \"description\": \"Product 1 desc\", \"price\": 10.34, \"categoryId\": 1}"))
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }

    @Test
    @Tag("DELETE")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void deleteProductByIDTest() throws Exception {

        mvc.perform(delete("/products/2"))
            .andExpect(status().isNoContent())
            .andReturn()
            .getResponse()
            .getContentAsString();
        // Add another assertion that checks the db and makes sure retired is set to true
    }

    @Test
    @Tag("DELETE")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void deleteProductByIDNotFoundTest() throws Exception {

        final var response = mvc.perform(delete("/products/10000")) 
            .andExpect(status().isNotFound())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }

    @Test
    @Tag("DELETE")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void deleteProductByIDInvalidTest() throws Exception {

        final var response = mvc.perform(delete("/products/invalid")) 
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        final String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        // This could be handled much better if the json was serialised.
        // Then we could point to the json field and sanatise using that.
        expect.serializer("json").toMatchSnapshot(sanitisedResponse);
    }
}
