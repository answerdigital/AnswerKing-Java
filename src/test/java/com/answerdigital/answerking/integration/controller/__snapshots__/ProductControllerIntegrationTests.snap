com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.deleteProductByIDInvalidTest=[
  ""
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.deleteProductByIDNotFoundTest=[
  "{\"type\":\"https://www.rfc-editor.org/rfc/rfc7231\",\"title\":\"Not Found Exception\",\"status\":404,\"detail\":\"One or more Products do not exist with ID(s) or name(s) [10000]. ERROR CODE - PRODUCT1\",\"instance\":\"/products/10000\",\"traceId\":\"{ Sanitised }\"}"
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.getAllProductsTest=[
  "[{\"id\":1,\"name\":\"Burger\",\"description\":\"300g Beef\",\"price\":6.69,\"category\":{\"id\":1,\"name\":\"CatOne\",\"description\":\"CatOne description\"},\"retired\":false},{\"id\":2,\"name\":\"Cola\",\"description\":\"500ml\",\"price\":2.99,\"category\":{\"id\":2,\"name\":\"CatTwo\",\"description\":\"CatTwo description\"},\"retired\":false},{\"id\":3,\"name\":\"Fries\",\"description\":\"Large Fries\",\"price\":2.99,\"category\":{\"id\":1,\"name\":\"CatOne\",\"description\":\"CatOne description\"},\"retired\":false}]"
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.getProductByIDInvalidTest=[
  ""
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.getProductByIDNotFoundTest=[
  "{\"type\":\"https://www.rfc-editor.org/rfc/rfc7231\",\"title\":\"Not Found Exception\",\"status\":404,\"detail\":\"One or more Products do not exist with ID(s) or name(s) [10000]. ERROR CODE - PRODUCT1\",\"instance\":\"/products/10000\",\"traceId\":\"{ Sanitised }\"}"
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.getProductByIDTest=[
  "{\"id\":1,\"name\":\"Burger\",\"description\":\"300g Beef\",\"price\":6.69,\"category\":{\"id\":1,\"name\":\"CatOne\",\"description\":\"CatOne description\"},\"retired\":false}"
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.postProductTest=[
  "{\"id\":4,\"name\":\"Kabab\",\"description\":\"Product 1 desc\",\"price\":10.34,\"category\":{\"id\":1,\"name\":\"CatOne\",\"description\":\"CatOne description\"},\"retired\":false}"
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.postProductTestInvalidBody=[
  ""
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.putProductByIDInvalidTest=[
  ""
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.putProductTest=[
  "{\"id\":2,\"name\":\"Kabab\",\"description\":\"Product 1 desc\",\"price\":10.34,\"category\":{\"id\":2,\"name\":\"CatTwo\",\"description\":\"CatTwo description\"},\"retired\":false}"
]


com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.putProductTestInvalidBody=[
  ""
]